newlib/libc/machine/arc64/strlen.S

```diff
@@ -0,0 +1,269 @@
+/*
+   Copyright (c) 2022, Synopsys, Inc. All rights reserved.
+
+   Redistribution and use in source and binary forms, with or without
+   modification, are permitted provided that the following conditions are met:
+
+   1) Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+   2) Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+   3) Neither the name of the Synopsys, Inc., nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+   THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+   AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+   ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+   LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+   CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+   POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#include <sys/asm.h>
+
+; Using 128-bit memory operations
+#if defined (__ARC64_M128__)
+
+; Code Brief (more info at the bottom):
+; Searches the provided string, 32 bytes at a time, using 128 bit loads
+; Finds the NULL bytes inside the loaded data
+; Analyzes the first NULL byte containing double word and calculates
+; size appropriately
+;
+; R0 const char* ptr (string to measure)
+; ret (R0):
+;		- unsigned (string size)
+;
+
+ENTRY (strlen)
+
+; Preserve r0 for size calculation when returning
+	MOVP	r13, r0
+	xorl	r6, r6, r6
+
+; Setup byte detector (more information bellow) [1]
+	vpack2wl	r8, NULL_32DT_1, NULL_32DT_1
+	asll	r9, r8, 7
+
+
+.L_4_8B_search:
+	lddl.ab	r2r3, [r13, +16]
```

Comment:
As far as I can see the only difference between having double load/store instructions and none is only in this place.
My suggestion is to use here an IFDEF and scrap the second copy of this algorithm found below.